### ISSUE TITLE
Poor performance casting to Literal with many values

### ISSUE THREAD

**Issue by tekumara** [microsoft/pyright#1817]
**Describe the bug**

The boto type stubs have introduced [InstanceType](https://github.com/vemel/boto3_stubs_docs/blob/master/mypy_boto3_ec2/literals.md#instancetype), a Literal with ~400 values. 

Casting to InstanceType takes ~60 secs on the command line. vscode consumes 100% CPU. 

**To Reproduce**

_test.py_
```python
from typing import cast
import boto3
from mypy_boto3_ec2.literals import InstanceType

def launch(instance_type:str) -> None:
    ec2_client = boto3.client("ec2")
    response = ec2_client.run_instances(MinCount=1, MaxCount=1, InstanceType=cast(InstanceType,instance_type))
```

```
$ pip install boto3==1.17.62 'boto3-stubs[ec2]==1.17.62.0'
$ time pyright test.py
No configuration file found.
No pyproject.toml file found.
stubPath /Users/tekumara/code/aec/typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
0 errors, 0 warnings, 0 infos 
Completed in 56.345sec
node_modules/.bin/pyright src/aec/command/test.py  56.37s user 0.65s system 100% cpu 56.550 total
```

**Expected behavior**

Type-checking should take < 1 sec.

**Additional context**
pyright 1.1.136
Pylance language server 2021.4.3 (pyright 320a0a2f)

**Comment by erictraut** [comment]
Thanks for the bug report. We were hitting an O(n*m) performance issue (where both n and m were roughly 400). I've added additional code to handle this case when both the source and destination union types contain mostly string or integer literals.

**Comment by erictraut** [comment]
This is included in pyright 1.1.137, which I just published. It will also be included in the next release of pylance.
